### PR TITLE
Correctly redact credentials when using x-oauth-basic 

### PR DESCRIPTION
**Pull request by mlarraz**
## What was the end-user or developer problem that led to this PR?
When setting credentials for Github Package Registry, Github allows 2 separate formats:

* `<username>:<token>`
* `<token>:x-oauth_basic`

In the latter case, Bundler fails to redact the OAuth token.

## What is your fix for the problem, implemented in this PR?
When redacting, check if the password is `x-oauth_basic`, in which case assume the username is the actual value to redact.

### CODE REVIEW

**Review by deivid-rodriguez — APPROVED** [review]
This looks good to me! Only comment/question is whether we should also redact the username when we get the form `https://<token>@github.com/owner/repo.git`, which is apparently also valid, according to https://github.blog/2012-09-21-easier-builds-and-deployments-using-git-over-https-and-oauth/.

Maybe @intuxicated also has some feedback since he has setup a private repository recently to reproduce another issue, and he's also working on a different credential redaction issue.

**Review by intuxicated — APPROVED** [review]
Thanks for doing this 🎉 .

My understanding of `x-oauth-basic` is that Github has deprecated it and to be honest I'm surprised it's still working.

**Comment by intuxicated** [comment]
@deivid-rodriguez After thinking about this for a while, I think it would be good to redact both username and password. I can't find a useful use-case in which exposing the username is necessary. 

If you agree, we need to apply the same behavior to `rubygems`.

File: bundler/lib/bundler/settings.rb
Comment (deivid-rodriguez) [reply bundler/lib/bundler/settings.rb]: After seeing the updated code, I'm actually not sure the new code is more clear. With the previous version, it was kind of clear that `x-oauth-basic` at the place of the password was an special case where the username should be treated as the redactable part. Even Github docs talk about

> You can simply use an OAuth token for the username and either a blank password or the string x-oauth-basic when cloning a repository.

Sorry @mlarraz for the back and forth, don't update this yet, if @intuxicated still feels `key:value` is better, we can stick to this version, it's no big deal.

**Comment by deivid-rodriguez** [comment]
I have doubts, I understand that it might not be necessary but that doesn't mean it should be redacted? Redacting credentials should be to prevent people from accidentally copy-pasting them to the open, but I don't think copy-pasting a username should be considered a security problem? It feels a bit like security by obscurity.

File: bundler/lib/bundler/settings.rb
Comment (intuxicated) [reply bundler/lib/bundler/settings.rb]: Yeah, you are right @deivid-rodriguez. I look into the ruby implementation of [URI::Generic#split_userinfo](https://github.com/ruby/ruby/blob/f51a6ace06cecd193bf00747a63809849e254050/lib/uri/generic.rb#L540-L548) and it's the same thing there, `username, password`. I think the only improvement here would be providing the integer `2` as the second argument to prevent unexpected behavior.

```ruby
irb(main)>  URI('http://user:password@ecample.com').userinfo.split(':')
=> ["user", "password"] # ok
irb(main)>  URI('http://user:password@ecample.com').userinfo.split(':', 2)
=> ["user", "password"] # ok
irb(main)>  URI('http://user:pass:word@ecample.com').userinfo.split(':')
=> ["user", "pass", "word"] # bad
irb(main)>  URI('http://user:pass:word@ecample.com').userinfo.split(':', 2)
=> ["user", "pass:word"] # prefect - cover both scenario

```

Comment (deivid-rodriguez) [reply bundler/lib/bundler/settings.rb]: Definitely :+1:

**Comment by intuxicated** [comment]
In that case, I would say let's leave it as it is. if there was a strong case for redacting the username in the future, we will fix it.

File: bundler/lib/bundler/settings.rb
Comment (intuxicated) [reply bundler/lib/bundler/settings.rb]: Sorry, @mlarraz. If you could drop the `8e88d08` from your PR, it would be awesome. Also, add the second argument with the value of `2` to make sure we are not breaking things.
Also, If you don't have the time, you can leave it to me, I'll fix it. Thanks 🥇

**Review by deivid-rodriguez — APPROVED** [review]
It looks good to me! :+1:

I'll merge once @intuxicated is happy too!

**Comment by deivid-rodriguez** [comment]
@intuxicated had actually already approved this, so merging!

**Comment by deivid-rodriguez** [comment]
Thanks so much @mlarraz!!

**Pull request merged**
